### PR TITLE
Pipeline 2.01

### PR DESCRIPTION
**Pull request by nbischoffcog**
1. Adds Object Detection and Flow
2. Adds a script to do both Object Detection and Flow from the same file
3. Archives old files in /deep_video_libs/old
4. Adds functional I3D Arch
5. Updates to README

### CODE REVIEW

File: deep_video_libs/I3D_arch.py [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+    x = MaxPooling3D((1,3,3), strides = (1,2,2), padding = 'same', name = 'MaxPool2d_2a_3x3')(x)
+    x = conv3d_layer(x, 64, 1,1,1, strides = (1,1,1), padding = 'same', name = 'Conv3d_2b_1x1')
+    x = conv3d_layer(x, 192, 3,3,3, strides = (1,1,1), padding = 'same', name = 'Conv3d_2c_3x3')
+
+    #MaxPooling
+    x = MaxPooling3D((1,3,3), strides = (1,2,2), padding = 'same', name = 'MaxPool2d_3a_3x3')(x)
+
+    #Mixed 3B
+
+    branch_0 = conv3d_layer(x, 64, 1,1,1, padding = 'same', name = 'Conv3d_3b_0a_1x1')
```
Comment (fgarberson) [inline deep_video_libs/I3D_arch.py]: I'm a bit confused why there are this many branches here?

File: deep_video_libs/I3D_arch.py [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+            else:
+                weights_path = '/rgb_inception_i3d_kinetics_only_tf_dim_ordering_tf_kernels_no_top.h5'
+
+        elif weights == WEIGHTS_NAME[1]: # flow_kinetics_only
+            if include_top:
+                weights_path = '/flow_inception_i3d_kinetics_only_tf_dim_ordering_tf_kernels.h5'
+            else:
+                weights_path = '/flow_inception_i3d_kinetics_only_tf_dim_ordering_tf_kernels_no_top.h5'
+
+        elif weights == WEIGHTS_NAME[2]:  # rgb_imagenet_and_kinetics
```
Comment (fgarberson) [inline deep_video_libs/I3D_arch.py]: I thought that optical flow got its own 3D CNN in I3D? Or is optical flow merged in at a later stage. The only effect of optical flow that I see here is a different weight file being loaded.

File: deep_video_libs/I3D_arch.py (hunk 1 above)
Comment (nbischoffcog) [reply deep_video_libs/I3D_arch.py]: The inception network (the backbone of I3D) has a number of branches, i can add the image of the architectures to the README file

File: deep_video_libs/I3D_arch.py [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+
+def top_of_model(classes = 400, n_gpus=1,train=True):
+    
+    if train:
+        model = inflated3d_inception(weights = 'rgb_kinetics_only',input_shape=(224,224,224,3), dropout_prob=0.25, endpoint_logit=False, classes=classes, n_gpus=n_gpus)
+#         img_input = Input(model.layers[-1].output)
+        x = conv3d_layer(model.layers[-1].output, classes, 1, 1, 1, padding='same', use_bias=True, use_activation_fn=False, use_bn=False, name='Conv3d_6a_1x1')
+        num_frames_remaining = int(x.shape[1])
+        x = Reshape((num_frames_remaining, classes))(x)
+        x = Lambda(lambda x: K.mean(x, axis=1, keepdims=False), output_shape=lambda s: (s[0], s[2]))(x)
```
Comment (fgarberson) [inline deep_video_libs/I3D_arch.py]: So when you include the top layer it wipes out the classes ...

File: deep_video_libs/I3D_arch.py (hunk 2 above)
Comment (nbischoffcog) [reply deep_video_libs/I3D_arch.py]: In the actual training, the logits of the rgb channel and flow channel will be added and a softmax applied to find the class

File: deep_video_libs/object_detection_and_calculate_flow.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+cnn_face_detector = dlib.cnn_face_detection_model_v1(face_crop_weights)
+###Load and customize the object detector from imageai
+detector = ObjectDetection()
+detector.setModelTypeAsYOLOv3()
+weights = os.path.join(DV_home , "weights/yolo.h5")
+detector.setModelPath(weights)
+detector.loadModel(detection_speed="fastest")
+custom = detector.CustomObjects(person=True)
+###Load the TVL1 model for flow detections
+TVL1 = DualTVL1()
```
Comment (fgarberson) [inline deep_video_libs/object_detection_and_calculate_flow.py]: This is the optical flow algorithm I3D uses, right?

Comment (nbischoffcog) [reply deep_video_libs/object_detection_and_calculate_flow.py]: yes

File: deep_video_libs/object_detection_and_calculate_flow.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+        curr_original = cv2.imread(img_file)
+        curr_rgb = cv2.cvtColor(curr_original, cv2.COLOR_BGR2RGB)
+        
+        if do_flow:
+            ##### do Flow Calculations and append to the full video flow
+            if curr_rgb is not None:
+                prev_gray = cv2.cvtColor(prev_rgb,cv2.COLOR_BGR2GRAY)
+                curr_gray = cv2.cvtColor(curr_rgb,cv2.COLOR_BGR2GRAY)
+
+                curr_flow = TVL1.calc(prev_gray, curr_gray, None)
```
Comment (fgarberson) [inline deep_video_libs/object_detection_and_calculate_flow.py]: Ok, so all flow calculations are based on single frame-pairs ...

File: deep_video_libs/object_detection_and_calculate_flow.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+                assert(curr_flow.dtype == np.float32)
+                # truncate [-20, 20]
+                curr_flow[curr_flow >= 20] = 20
```
Comment (fgarberson) [inline deep_video_libs/object_detection_and_calculate_flow.py]: I don't suppose that we have any idea how often we hit this limit? ... Or if the same limit was also used consistently with I3D ... ?

Comment (nbischoffcog) [reply deep_video_libs/object_detection_and_calculate_flow.py]: I dont know, is it worth the investigation?

File: deep_video_libs/object_detection_and_calculate_flow.py (hunk 5 above)
Comment (nbischoffcog) [reply deep_video_libs/object_detection_and_calculate_flow.py]: Yes that is correct

File: deep_video_libs/object_detection_and_calculate_flow.py (hunk 6 above)
Comment (fgarberson) [reply deep_video_libs/object_detection_and_calculate_flow.py]: I think it's worth checking if we are hitting it often. If it's less than 5% of the time I wouldn't worry about it. If it's more than 10% of the time it could be a big deal.
